File: NISTv4.2.1/DNAnexusApplets/nist-integration-v4.2.1-anyref/resources/usr/bin/preprocess_combine_vcfs.pl
#To load following perl modules
use strict;
use warnings;
use Getopt::Long;

#Input parameters
my (@CALLSETTABLE, $FILEPATH);
my ($verbose);

#Declare variables
GetOptions(
	'cstable=s{1,}' => \@CALLSETTABLE,
	'p|filepath:s' => \$FILEPATH,
	"h|help|?" => \&usage);

#To check input files
my $CALLSETTABLE = $CALLSETTABLE[0];
if(defined($FILEPATH)){$FILEPATH = $FILEPATH} else {print usage();die "\n\nPlease specify the path of the vcf and bed files.\n\n"}

#To define usage of the preprocess_combine_vcfs
print "Usage = perl preprocess_combine_vcfs.pl -cstable $CALLSETTABLE ";
print "-p $FILEPATH \n\n";

sub usage {
	print "\nusage: perl preprocess_combine_vcfs.pl -cstable <txt> -p <FILEPATH> \n";
	print "\t-v \t\t\tThe input Tab-delimited text file containing columns Platform, Dataset, Callset, vcfAll.vcf.gz, callableBed.bed, and annotationsFile \n";
	print "\t-p, --filepath\t\t\t\tPath where vcf and bed files are located, ending in \\.\n";
	exit 1;
}

open (CALLSETTABLE, "<", $CALLSETTABLE) or die "Can't open the input table file. Please check if the file exists\n\n";

my $m = 1; #vcf number in order in callsettable
my $vcfs=""; #output variable that is printed that will contain names of processed vcfs for integration
my @line_head = (); #Callset table header line, which contains bed file names that may or may not be excluded from the callable regions of each callset

#loop through each line in callset table
while (my $line = <CALLSETTABLE>) {
	my $vcfno="$m";
	if ($m<10) {
		$vcfno="0$m";
	}
	chomp($line); #removes new line char
	
	my @line_data = split("\t", $line);
    
    if($line =~ m/^\#/) { #get header line and go to next line
        @line_head = split("\t", $line);
        next;
    }
    
    my $vcfstart=""; #begininning of vcf file name
    if($line_data[3] =~ /(.*)\.vcf.gz/) { $vcfstart=$1; } else { print "vcfAll does not end in .vcf.gz in cstable\n";exit 1; }
    
    #unzip vcf, keep first 10 columns, removes hom ref lines, removes chr prefix, breaks into "primitive" SNPs and indels, and sorts
    `gunzip -c ${FILEPATH}$line_data[3] | cut -f1-10 | grep -v '0/0' | grep -v '0|0' | vcfallelicprimitives -k -g | perl vcfsorter.pl genome.dict - > ${FILEPATH}${vcfstart}_nohomref.vcf`;
    
    #open formatted vcf from above
    open (VCFIN, "<", "${FILEPATH}${vcfstart}_nohomref.vcf") or die "Can't open the vcf ${FILEPATH}${vcfstart}_nohomref.vcf. Please check if the file exists\n\n";
	#open output vcf with additional formatting
	open (OUTINDIV, ">", "${FILEPATH}VCF${vcfno}_${vcfstart}_nohomref_samplename.vcf") or die $!;
	#open output bed that will be used to remove regions with duplicate vcf lines (mostly due to a bug in freebayes parallel that outputs multiple vcf lines at edge of parallelized regions
	open (OUTDUPBED, ">", "${FILEPATH}VCF${vcfno}_${vcfstart}_nohomref_dup.bed") or die $!;
    
    my $lineprev=<VCFIN>;
    chomp($lineprev);
    my @line_data1prev = split("\t", $lineprev); #read in first line of vcf and put in @line_data1prev
    my $duppos=0;
    my $anydup=0;
    while (my $line1 = <VCFIN>) {
		chomp($line1);
		my @line_data1 = split("\t", $line1);
	
		if($lineprev =~ m/^\#CHROM/) { #change sample name so that they are ordered in the sequence of callsettable when using vcfcombine
			print OUTINDIV "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tVCF${vcfno}_$line_data[2]\n";
			#print  "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tVCF${m}_$line_data[2]\n";
		} elsif ($lineprev =~ m/^\#/) {
			print OUTINDIV "$lineprev\n";
		} elsif ( $line_data1prev[1]==$line_data1[1] ) {
			$duppos=1; #duplicate position, so add line to bed file that excludes +-50bp around these vcf lines to avoid vcfcombine bug that ignores all remaining variants
			$anydup=1;
			my $bed1=$line_data1[1]-50;
			my $bed2=$line_data1[1]+50;
			print OUTDUPBED "$line_data1[0]\t$bed1\t$bed2\n";
			print "Duplicated vcf lines: $lineprev\n$line1\n$line_data1[0]\t$bed1\t$bed2\n";
		} elsif ($duppos==1) { #don't print the previous line if it was a duplicate
			$duppos=0;
		} else { #otherwise, print the previous line
			print OUTINDIV "$lineprev\n";
		}
		$lineprev=$line1;
		@line_data1prev = split("\t", $line1);
	}
	print OUTINDIV "$lineprev\n";
	close VCFIN;
	close OUTINDIV;
	close OUTDUPBED;
	
	#bgzip and index vcf output above
	`bgzip -c ${FILEPATH}VCF${vcfno}_${vcfstart}_nohomref_samplename.vcf > ${FILEPATH}VCF${vcfno}_${vcfstart}_nohomref_samplename.vcf.gz`;
	`tabix -p vcf ${FILEPATH}VCF${vcfno}_${vcfstart}_nohomref_samplename.vcf.gz`;
  	$vcfs="$vcfs ${FILEPATH}VCF${vcfno}_${vcfstart}_nohomref_samplename.vcf.gz";
	
	if (!($line_data[4] eq "none")) { #skip if no callable regions
		#Find lines with CALLABLE in column 4 (if callableloci bed file) or only 3 columns in bed file (assume all are callable in this case)
		if (`awk '{ print NF; exit }' ${FILEPATH}$line_data[4]` > 3) {
			if ($anydup==1) {
				`grep CALLABLE ${FILEPATH}$line_data[4] | cut -f1-3 | subtractBed -a stdin -b ${FILEPATH}VCF${vcfno}_${vcfstart}_nohomref_dup.bed > ${FILEPATH}${vcfstart}_callable_tmp_1.bed`;
			} else {
				`grep CALLABLE ${FILEPATH}$line_data[4] > ${FILEPATH}${vcfstart}_callable_tmp_1.bed`;
			}
		} else {
			if ($anydup==1) {
				`cut -f1-3 ${FILEPATH}$line_data[4] | subtractBed -a stdin -b ${FILEPATH}VCF${vcfno}_${vcfstart}_nohomref_dup.bed > ${FILEPATH}${vcfstart}_callable_tmp_1.bed`;
			} else {
				`mv ${FILEPATH}$line_data[4] ${FILEPATH}${vcfstart}_callable_tmp_1.bed`;
			}
		}
	
		#in v3.3, now subtract general difficult bed files from each callset's callable bed only when specified in callsettable
		my $c=6; #start with 7th column, which is the first bed file to exclude
		my $d=1;
		while ($c<=$#line_head) {
			if ($line_data[$c]==1) {
				`subtractBed -a ${FILEPATH}${vcfstart}_callable_tmp_${d}.bed -b ${line_head[$c]} > ${FILEPATH}${vcfstart}_callable_tmp_${c}.bed`;
				$d=$c;
			}
			$c++;
		}
		`awk 'BEGIN {FS = OFS = "\t"} {print \$1 "\t" \$2 "\t" \$3 "\t" "CS_$line_data[2]_callable"}' ${FILEPATH}${vcfstart}_callable_tmp_${d}.bed > ${FILEPATH}${vcfstart}_callable_processed.bed`;
		`rm ${FILEPATH}${vcfstart}_callable_tmp*.bed `;
	}
	$m=$m+1;

}
print "$vcfs\n";
#`vcfs="$vcfs"`;  #set environment variable that will contain names of processed vcfs for integration
close CALLSETTABLE;

#Add column to general difficult bed files to add annotations to union vcf
my $c=6; #start with 7th column, which is the first bed file to exclude
while ($c<=$#line_head) {
	my $bedname=$line_head[$c];
	if ($line_head[$c] =~ /(.*).bed.gz/) { 
		$bedname=$1; 
		`gunzip -c ${line_head[$c]} | awk 'BEGIN {FS = OFS = "\t"} {print \$1 "\t" \$2 "\t" \$3 "\t" "$bedname"}' > ${bedname}_diffbedannotcol.bed`;
	} elsif ($line_head[$c] =~ /(.*).bed/) { 
		$bedname=$1; 
		`awk 'BEGIN {FS = OFS = "\t"} {print \$1 "\t" \$2 "\t" \$3 "\t" "$bedname"}' ${line_head[$c]} > ${bedname}_diffbedannotcol.bed`;
	} 
	$c++;
}


 

File: NISTv4.2.1/DNAnexusApplets/nist-integration-v4.2.1-anyref/resources/home/dnanexus/preprocess_combine_vcfs.py
import argparse
import subprocess
import os
import glob
import re

class PreProcessCombine(object):
	parser = None
	args = None
	path = None
	cstable_file = None
	cstable_file_lines = None
	cstable_file_header = None
	output_file_lines = None
	m = None
	vcfs = None
	cstable_line_head = None
	vcf_no_string = None
	line = None
	line_data = None
	vcf_prefix = None
	vcfin_path = None
	vcfin_file = None
			
	outindiv_path = None
	outindiv_file = None
	
	outdupbed_path = None
	outdupbed_file = None
	
	vcfin_lines = None
	duplicate_position = None
	anyduplicate = None
	vcf_previous_line = None
	line_data_previous = None
	split_vcf_line = None

	duplicate_position = None
	anyduplicate = None
	bed1 = None
	bed2 = None
	temp_callable_loci_bed_file = None
	temp_callable_loci_cols = None
	bedname = None

	def __init__(self, cstable, p):
		# self.parser = argparse.ArgumentParser(description='Process and combine VCF files.')

		# self.parser.add_argument('-cstable', metavar='txt', help='The input Tab-delimited text file containing columns Platform, Dataset, Callset, vcfAll.vcf.gz, callableBed.bed, and annotationsFile')
		# self.parser.add_argument('-p', metavar='FILEPATH', help='Path where vcf and bed files are located')

		
		self.path = p
		self.cstable_file = open(cstable, "r")

		self.cstable_file_lines = self.cstable_file.readlines()
		self.cstable_file_header = None
		self.output_file_lines = None
		self.m = None
		self.vcfs = None
		self.cstable_line_head = None
		self.vcf_no_string = None
		self.vcf_prefix = None
		self.vcfin_path = None
		self.vcfin_file = None
		
		self.outindiv_path = None
		self.outindiv_file = None
				
		self.outdupbed_path = None
		self.outdupbed_file = None
				
		self.vcfin_lines = None
		self.duplicate_position = None
		self.anyduplicate = None
		self.vcf_previous_line = None
		self.line_data_previous = None
		self.split_vcf_line = None

		self.duplicate_position = None
		self.anyduplicate = None
		self.bed1 = None
		self.bed2 = None
		self.temp_callable_loci_bed_file = None
		self.temp_callable_loci_cols = None
		self.bedname = None

	def process_callset_table(self):
		self.cstable_file_header = ""
		self.output_file_lines = []
		self.m = 1
		self.vcfs = ""
		self.cstable_line_head = []
		for line in self.cstable_file_lines:
			
			self.vcf_no_string = str(self.m)
			if self.m < 10:
				self.vcf_no_string = "0" + str(self.m)
		
			line = line.rstrip("\n")
			line_data = line.split("\t")
		
			if re.search("^\#", line):
				self.cstable_file_header = line
				self.cstable_line_head = self.cstable_file_header.split("\t")
				continue	

			if re.search("(.*)\.vcf.gz", line_data[3]):
				self.vcf_prefix = re.split("(.*)\.vcf.gz", line_data[3])[1]
			else:
				#print("vcfAll does not end in .vcf.gz")
				exit(1)
			#print("gunzip -c " + self.path + line_data[3] + " | cut -f1-10 | grep -v '0/0' | grep -v '0|0' | vcfallelicprimitives -k -g | perl vcfsorter.pl genome.dict - > " + self.path + self.vcf_prefix + "_nohomref.vcf")
			subprocess.check_call("gunzip -c " + self.path + line_data[3] + " | cut -f1-10 | grep -v '0/0' | grep -v '0|0' | vcfallelicprimitives -k -g | perl vcfsorter.pl genome.dict - > " + self.path + self.vcf_prefix + "_nohomref.vcf", shell=True)
			
			self.vcfin_path = self.path + self.vcf_prefix + "_nohomref.vcf"
			#print(self.vcfin_path)
			self.vcfin_file = open(self.vcfin_path, "r")
			
			self.outindiv_path = self.path + "VCF" + self.vcf_no_string + "_" + self.vcf_prefix + "_nohomref_samplename.vcf"
			#print(self.outindiv_path)
			self.outindiv_file = open(self.outindiv_path, "w+")
			
			self.outdupbed_path = self.path + "VCF" + self.vcf_no_string + "_" + self.vcf_prefix + "_nohomref_dup.bed"
			#print(self.outdupbed_path)
			self.outdupbed_file = open(self.outdupbed_path, "w+")
			
			self.vcfin_lines = self.vcfin_file.readlines()
			# #print(self.vcfin_lines)
			self.duplicate_position = 0
			self.anyduplicate = 0 
			self.vcf_previous_line = self.vcfin_lines[0].rstrip("\n")
			self.line_data_previous = self.vcf_previous_line.split("\t")
			
			for i in range(1,len(self.vcfin_lines)):
				self.vcfin_line = self.vcfin_lines[i].rstrip("\n")
				self.split_vcf_line = self.vcfin_line.split("\t")
				if self.vcf_previous_line.startswith('#CHROM'):
					self.outindiv_file.write("#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tVCF" + self.vcf_no_string + "_"+ self.split_vcf_line[2] + "\n")
				elif self.vcf_previous_line.startswith('#'):
					self.outindiv_file.write(self.vcf_previous_line + "\n")
				elif self.line_data_previous[1] == self.split_vcf_line[1]:
					self.duplicate_position = 1
					self.anyduplicate = 1
					self.bed1 = int(self.split_vcf_line[1]) - 50
					self.bed2 = int(self.split_vcf_line[1]) + 50
					#print("Duplicated vcf lines: " + self.vcf_previous_line + "\n" + self.vcfin_line + self.splint_vcf_line[0] + "\t" + str(self.bed1) + "\t" + str(self.bed2) + "\n")
					self.outdupbed_file.write(self.split_vcf_line[0]+"\t"+str(self.bed1)+"\t"+str(self.bed2)+"\n")
				elif self.duplicate_position == 1:
					self.duplicate_position = 0
				else:
					self.outindiv_file.write(self.vcf_previous_line + "\n")

				self.vcf_previous_line = self.vcfin_line
				self.line_data_previous = self.vcfin_line.split("\t")

			self.outindiv_file.write(self.vcf_previous_line + "\n")
			self.vcfin_file.close()
			self.outindiv_file.close() 
			self.outdupbed_file.close() 

			subprocess.check_call("bgzip -c " + self.path + "VCF" + self.vcf_no_string + "_" + self.vcf_prefix + "_nohomref_samplename.vcf > " + self.path + "VCF" + self.vcf_no_string + "_" + self.vcf_prefix + "_nohomref_samplename.vcf.gz", shell=True) 
			subprocess.check_call("tabix -p vcf " + self.path + "VCF" + self.vcf_no_string + "_" + self.vcf_prefix + "_nohomref_samplename.vcf.gz", shell=True)

			self.vcfs = self.vcfs + " " + self.path + "VCF" + self.vcf_no_string + "_" + self.vcf_prefix + "_nohomref_samplename.vcf.gz"	
			#print("any duplicate: " + str(self.anyduplicate))
			if not(line_data[4] == "none"):
				self.temp_callable_loci_bed_file = open(self.path+line_data[4], "r").readlines()
				self.temp_callable_loci_cols = self.temp_callable_loci_bed_file[0].split("\t")
				if len(self.temp_callable_loci_cols) > 3: 
					if self.anyduplicate == 1:
						subprocess.check_call("grep CALLABLE " + self.path + line_data[4] + " | cut -f1-3 | subtractBed -a stdin -b " + self.path + "VCF" + self.vcf_no_string + "_" + self.vcf_prefix + "_nohomref_dup.bed > " + self.path + self.vcf_prefix + "_callable_tmp_1.bed", shell=True) 
					else: 
						subprocess.check_call("grep CALLABLE " + self.path + line_data[4] + " > " + self.path + self.vcf_prefix + "_callable_tmp_1.bed", shell=True)
				else:
					if self.anyduplicate == 1:
						subprocess.check_call("cut -f1-3 " + self.path + line_data[4] + " | subtractBed -a stdin -b " + self.path + "VCF" + self.vcf_no_string + "_" + self.vcf_prefix + "_nohomref_dup.bed > " + self.path + self.vcf_prefix + "_callable_tmp_1.bed", shell=True)
					else:
						subprocess.check_call("cp " + self.path + line_data[4] + " " + self.path + self.vcf_prefix + "_callable_tmp_1.bed",shell=True)
				
				c = 6
				d = 1
				while c < len(self.cstable_line_head):
					if line_data[c] == "1":
						subprocess.check_call("subtractBed -a " + self.path + self.vcf_prefix + "_callable_tmp_" + str(d) + ".bed -b " + self.cstable_line_head[c] + " > " + self.path + self.vcf_prefix + "_callable_tmp_" + str(c) + ".bed", shell=True)
						d = c
					c = c + 1
				subprocess.check_call("awk \'BEGIN {FS = OFS = \"\t\"} {print $1 \"\t\" $2 \"\t\" $3 \"\t\" \"CS_"+ line_data[2] + "_callable\"}\' " + self.path + self.vcf_prefix + "_callable_tmp_" + str(d) + ".bed > " + self.path + self.vcf_prefix + "_callable_processed.bed", shell=True) 
			self.m = self.m + 1

		#print(self.vcfs + "\n")

		self.cstable_file.close()

	def add_difficult_annotation(self):
		c = 6
		while(c < len(self.cstable_line_head)):
			bedname = self.cstable_line_head[c]
			if re.search("(.*).bed.gz", self.cstable_line_head[c]):
				bedname = re.split("(.*).bed.gz", self.cstable_line_head[c])[1]
				#print(bedname)
				#print("gunzip -c " + self.cstable_line_head[c] + " | awk \'BEGIN {FS = OFS = \"\t\"} {#print $1 \"\t\" $2 \"\t\" $3 \"\t\" \"" + bedname + "\"}\' > " + bedname + "_diffbedannotcol.bed")
				subprocess.check_call("gunzip -c " + self.cstable_line_head[c] + " | awk \'BEGIN {FS = OFS = \"\t\"} {print $1 \"\t\" $2  \"\t\" $3 \"\t\" \"" + bedname + "\"}\' > " + bedname + "_diffbedannotcol.bed", shell=True)
			elif re.search("(.*).bed", self.cstable_line_head[c]):
				bedname = re.split("(.*).bed", self.cstable_line_head[c])[1]
				#print(bedname)
				#print("awk \'BEGIN {FS = OFS = \"\t\"} {#print $1 \"\t\" $2 \"\t\" $3 \"\t\" \"" + bedname + "\"}\' " + self.cstable_line_head[c] + " > " + bedname + "_diffbedannotcol.bed")
				subprocess.check_call("awk \'BEGIN {FS = OFS = \"\t\"} {print $1 \"\t\" $2 \"\t\" $3 \"\t\" \"" + bedname + "\"}\' " + self.cstable_line_head[c] + " > " + bedname + "_diffbedannotcol.bed", shell=True) 
			c = c + 1

	def preprocess(self):
		self.process_callset_table()
		self.add_difficult_annotation()
